contracts/root/withdrawManager/WithdrawManager.sol

```diff
@@ -5,260 +5,121 @@ import { ERC721 } from "openzeppelin-solidity/contracts/token/ERC721/ERC721.sol"
+    require(
+      MerklePatriciaProof.verify(
+        referenceTxData[offset + 6].toBytes(), // receipt
+        branchMask,
+        referenceTxData[offset + 7].toBytes(), // receiptProof
+        bytes32(referenceTxData[offset + 5].toUint()) // receiptsRoot
+      ),
+      "INVALID_RECEIPT_MERKLE_PROOF"
     );
-  }
 
-  function createExitQueue(address _token)
-    external
-  {
-    require(msg.sender == address(registry), "UNAUTHORIZED_REGISTRY_ONLY");
-    exitsQueues[_token] = address(new PriorityQueue());
-  }
+    if(verifyTxInclusion) {
+      require(
+        MerklePatriciaProof.verify(
+          referenceTxData[offset + 10].toBytes(), // tx
+          branchMask,
+          referenceTxData[offset + 11].toBytes(), // txProof
+          bytes32(referenceTxData[offset + 4].toUint()) // txRoot
+        ),
+        "INVALID_TX_MERKLE_PROOF"
+      );
+    }
 
-  function _withdraw(
-    PlasmaExit memory exitObject,
-    uint256 headerNumber,
-    bytes memory withdrawBlockProof,
-    uint256 withdrawBlockNumber,
-    uint256 withdrawBlockTime,
-    bytes32 withdrawBlockTxRoot,
-    bytes32 withdrawBlockReceiptRoot,
-    bytes memory path,
-    uint8 oIndex)
-    internal
-  {
     uint256 startBlock;
     bytes32 headerRoot;
-
     // @todo a function to return just root and startBlock might save gas
     (headerRoot, startBlock,,,) = rootChain.headerBlocks(headerNumber);
 
+    uint256 blockNumber = referenceTxData[offset + 2].toUint();
     require(
-      keccak256(abi.encodePacked(withdrawBlockNumber, withdrawBlockTime, withdrawBlockTxRoot, withdrawBlockReceiptRoot))
-        .checkMembership(withdrawBlockNumber - startBlock, headerRoot, withdrawBlockProof),
+      keccak256(abi.encodePacked(
+        blockNumber,
+        referenceTxData[offset + 3].toUint(), // blockTime
+        bytes32(referenceTxData[offset + 4].toUint()), // txRoot
+        bytes32(referenceTxData[offset + 5].toUint()) // receiptRoot
+      )).checkMembership(blockNumber - startBlock, headerRoot, referenceTxData[offset + 1].toBytes() /* blockProof */),
       "WITHDRAW_BLOCK_NOT_A_PART_OF_SUBMITTED_HEADER"
     );
 
-    uint256 _exitId = (
+    age = (
       headerNumber * HEADER_BLOCK_NUMBER_WEIGHT +
-      withdrawBlockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
-      path.toRlpItem().toBytes().toRlpItem().toUint() * 100000 +
-      oIndex
+      blockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
+      branchMask.toRlpItem().toBytes().toRlpItem().toUint() * BRANCH_MASK_WEIGHT
     );
+  }
+
+  function createExitQueue(address token)
+    external
+  {
+    require(msg.sender == address(registry), "UNAUTHORIZED_REGISTRY_ONLY");
+    exitsQueues[token] = address(new PriorityQueue());
+  }
 
-    _addExitToQueue(exitObject, _exitId, withdrawBlockTime);
+  modifier isPredicateAuthorized(address rootToken) {
+    (Registry.Type _type) = registry.predicates(msg.sender);
+    if (_type == Registry.Type.ERC20) {
```

Comment:
why call `registry.isERC721(rootToken)` when we can do short-circuit in single require statement and `Registry.Type.ERC20/ERC721` also Enum is internally uint, we can benefit from that in same `require` short-circuit logic.

Reply:
Didn't understand 😅 Can you write a code snippet?